### PR TITLE
Sockets - Evelynn

### PR DESCRIPTION
**Pull request by evelynnkaplan**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | This both prevents the program from saving unexpected input, and informs the user that the input they've given is unexpected and not allowed.
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? | These methods are used to search for a particular instance among all instances of a class. Instance methods are to be called on a particular instance, and finding an instance by calling on a random instance doesn't make sense.
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | This relationship is one-to-many in that one Customer can belong to many Orders. But each Order only has one instance of a Customer. This is different than the Solar System project, even though Solar System was also one-to-many. Many instances of a Planet belonged to the SolarSystem.
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | In both, the ID number of the Order is the focus of the Order class, and the ID number of the Customer associated with the Order comes secondary (because there is no "find all orders belonging to one customer" method. However, if this method of looking up orders was created instead of the current Order#find method (which looks up an order by order ID, not customer ID), the relationship in the code would show that a Customer can belong to many orders. The CSV makes it appear as though the relationship might be one-to-one until you notice that customer IDs are repeated in some places.
Did the presence of automated tests change the way you thought about the problem? How? | Yes, because it encouraged raising Argument Errors for certain edge cases.
CS Fun: If there are `C` customers listed in `customers.csv` and `R` orders listed in `orders.csv`, what is the time complexity of calling `Customer.find`? What is the time complexity of `Order.find`? |

### CODE REVIEW

File: lib/customer.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  def initialize(cust_id, cust_email, cust_address)
+    @cust_id = cust_id
+    @cust_email = cust_email
+    @cust_address = cust_address
+  end
+
+  def id
+    return @cust_id
+  end
```
Comment (droberts-sea) [inline lib/customer.rb:15]: You don't need to hand-write these getter methods if you use the `attr_reader` and `attr_accessor` helpers above.

File: lib/customer.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        customers << customer
+      end
+    end
+    return customers
+  end
+
+  def self.find(id)
+    customers = self.all
+    customers.each do |customer|
+      if customer.id == id
```
Comment (droberts-sea) [inline lib/customer.rb:44]: Is there a builtin enumerable method that might help clarify this code?

File: lib/order.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+  def initialize(id, products, customer, status = :pending)
+    @id = id
+    @products = products
+    @customer = customer
+
+    valid_statuses = %i[pending paid processing shipped complete]
+    if !valid_statuses.include?(status)
+      raise ArgumentError, "Only order statuses are #{valid_statuses}."
+    end
```
Comment (droberts-sea) [inline lib/order.rb:16]: This list of statuses would be a great thing to store in a constant.

File: lib/order.rb [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  def self.all
+    orders = []
+    CSV.open("data/orders.csv") do |file|
+      file.each do |row|
+        customer = Customer.find(row[2].to_i)
+        items_string = row[1]
+        items_array = items_string.split(";")
+        products = {}
+        items_array.each do |item|
+          product_array = item.split(":")
```
Comment (droberts-sea) [inline lib/order.rb:57]: Good work getting this tricky order-parsing logic figured out!

**Comment by droberts-sea** [comment]
# Grocery Store
## What We're Looking For

Feature | Feedback
--- | ---
**Baseline** |
Answered comprehension questions | yes
Used Git Regularly | your commit frequency is good, but I would like to see more descriptive commit messages. In particular, instead of referencing the different waves, you should describe what code you've written.
**Wave 1** |
All provided tests pass | yes
Using the appropriate `attr_` for instance variables | yes, but see inline
**Wave 2** |
All stubbed tests are implemented fully and pass | yes
Used `CSV` library only in `.all` (not in `.find`) | yes
Appropriately parses the product data from CSV file in `Order.all` | yes
`Order.all` calls `Customer.find` to set up the composition relation | yes
**Additional Notes** | Good job overall! Your code is well-organized and easy to read, and does a good job of solving the problem at hand. I've left a few inline comments for you to review, but it's clear that the learning goals for this assignment were met. Keep up the hard work!
